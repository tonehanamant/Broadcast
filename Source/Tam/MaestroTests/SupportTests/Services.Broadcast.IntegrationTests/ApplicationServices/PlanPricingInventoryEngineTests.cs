using ApprovalTests;
using ApprovalTests.Reporters;
using IntegrationTests.Common;
using Newtonsoft.Json;
using NUnit.Framework;
using Services.Broadcast.BusinessEngines;
using Services.Broadcast.Entities.Plan.Pricing;
using Services.Broadcast.IntegrationTests.Helpers;
using Services.Broadcast.IntegrationTests.Stubs;
using Services.Broadcast.Repositories;
using Tam.Maestro.Common.DataLayer;
using static Services.Broadcast.Entities.ProposalProgramDto;

namespace Services.Broadcast.IntegrationTests.ApplicationServices
{
    [TestFixture]
    public class PlanPricingInventoryEngineTests
    {
        private readonly IPlanPricingInventoryEngine _PlanPricingInventoryEngine;
        private readonly IPlanRepository _PlanRepository;
        private readonly InventoryFileTestHelper _InventoryFileTestHelper;

        public PlanPricingInventoryEngineTests()
        {
            _PlanPricingInventoryEngine = IntegrationTestApplicationServiceFactory.GetApplicationService<IPlanPricingInventoryEngine>();
            _PlanRepository = IntegrationTestApplicationServiceFactory.BroadcastDataRepositoryFactory.GetDataRepository<IPlanRepository>();
            _InventoryFileTestHelper = new InventoryFileTestHelper();
        }

        [Test]
        [UseReporter(typeof(DiffReporter))]
        public void RunNoDataTest()
        {
            using (new TransactionScopeWrapper())
            {
                var plan = _PlanRepository.GetPlan(1196);
                var result = _PlanPricingInventoryEngine.GetInventoryForPlan(plan, new PlanPricingInventoryEngine.ProgramInventoryOptionalParametersDto());

                Approvals.Verify(IntegrationTestHelper.ConvertToJson(result));
            }
        }

        [Test]
        [UseReporter(typeof(DiffReporter))]
        public void RunTest()
        {
            using (new TransactionScopeWrapper())
            {
                _InventoryFileTestHelper.UploadProprietaryInventoryFile(
                    "PricingModel_Barter.xlsx", 
                    processInventoryRatings: true, 
                    processInventoryProgramsJob: true);

                _InventoryFileTestHelper.UploadProprietaryInventoryFile(
                    "PricingModel_OAndO.xlsx", 
                    processInventoryRatings: true, 
                    processInventoryProgramsJob: true);

                var plan = _PlanRepository.GetPlan(1198);
                var result = _PlanPricingInventoryEngine.GetInventoryForPlan(plan, new PlanPricingInventoryEngine.ProgramInventoryOptionalParametersDto());

                var jsonResolver = new IgnorableSerializerContractResolver();
                jsonResolver.Ignore(typeof(PlanPricingInventoryProgram), "ManifestId");
                jsonResolver.Ignore(typeof(ManifestDaypartDto), "Id");
                var jsonSettings = new JsonSerializerSettings()
                {
                    ReferenceLoopHandling = ReferenceLoopHandling.Ignore,
                    ContractResolver = jsonResolver
                };
                var json = IntegrationTestHelper.ConvertToJson(result, jsonSettings);

                Approvals.Verify(json);
            }
        }

        [Test]
        [UseReporter(typeof(DiffReporter))]
        public void RunTest_DoesNotShowOpenMarketInventory_WhenSwitchIsOff()
        {
            using (new TransactionScopeWrapper())
            {
                StubbedConfigurationWebApiClient.RunTimeParameters["EnableOpenMarketInventoryForPricingModel"] = "False";

                try
                {
                    _InventoryFileTestHelper.UploadProprietaryInventoryFile(
                        "PricingModel_OAndO.xlsx",
                        processInventoryRatings: true,
                        processInventoryProgramsJob: true);

                    var plan = _PlanRepository.GetPlan(1198);
                    var result = _PlanPricingInventoryEngine.GetInventoryForPlan(plan, new PlanPricingInventoryEngine.ProgramInventoryOptionalParametersDto());

                    var jsonResolver = new IgnorableSerializerContractResolver();
                    jsonResolver.Ignore(typeof(PlanPricingInventoryProgram), "ManifestId");
                    jsonResolver.Ignore(typeof(ManifestDaypartDto), "Id");
                    var jsonSettings = new JsonSerializerSettings()
                    {
                        ReferenceLoopHandling = ReferenceLoopHandling.Ignore,
                        ContractResolver = jsonResolver
                    };
                    var json = IntegrationTestHelper.ConvertToJson(result, jsonSettings);

                    Approvals.Verify(json);
                }
                finally
                {
                    StubbedConfigurationWebApiClient.RunTimeParameters.Clear();
                }
            }
        }

        [Test]
        [UseReporter(typeof(DiffReporter))]
        public void RunWithHiatusDayTest()
        {
            using (new TransactionScopeWrapper())
            {
                var plan = _PlanRepository.GetPlan(1199);
                var result = _PlanPricingInventoryEngine.GetInventoryForPlan(plan, new PlanPricingInventoryEngine.ProgramInventoryOptionalParametersDto());

                Approvals.Verify(IntegrationTestHelper.ConvertToJson(result));
            }
        }

        [Test]
        [UseReporter(typeof(DiffReporter))]
        public void RunSpotLengthTest()
        {
            using (new TransactionScopeWrapper())
            {
                var plan = _PlanRepository.GetPlan(1197);
                var result = _PlanPricingInventoryEngine.GetInventoryForPlan(plan, new PlanPricingInventoryEngine.ProgramInventoryOptionalParametersDto());

                Approvals.Verify(IntegrationTestHelper.ConvertToJson(result));
            }
        }

        [Test]
        [UseReporter(typeof(DiffReporter))]
        public void RunWithProgramWithMultipleAudiencesTest()
        {
            using (new TransactionScopeWrapper())
            {
                var plan = _PlanRepository.GetPlan(1200);
                var result = _PlanPricingInventoryEngine.GetInventoryForPlan(plan, new PlanPricingInventoryEngine.ProgramInventoryOptionalParametersDto());

                Approvals.Verify(IntegrationTestHelper.ConvertToJson(result));
            }
        }
    }
}
