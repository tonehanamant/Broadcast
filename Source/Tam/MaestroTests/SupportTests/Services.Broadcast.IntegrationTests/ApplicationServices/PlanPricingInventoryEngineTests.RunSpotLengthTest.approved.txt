[
  {
    "InventorySource": "Open Market",
    "InventorySourceType": "Open Market",
    "ManifestAudiences": [
      {
        "AudienceId": 31,
        "Impressions": 34500.0,
        "IsReference": true
      },
      {
        "AudienceId": 33,
        "Impressions": 22700.0,
        "IsReference": true
      },
      {
        "AudienceId": 37,
        "Impressions": 26200.0,
        "IsReference": true
      },
      {
        "AudienceId": 39,
        "Impressions": 20000.0,
        "IsReference": true
      },
      {
        "AudienceId": 40,
        "Impressions": 27900.0,
        "IsReference": true
      }
    ],
    "ManifestDayparts": [
      {
        "DaypartId": 32941,
        "Id": 358772,
        "ProgramName": "FRIENDS"
      }
    ],
    "ManifestId": 287146,
    "MediaWeekIds": [
      771,
      772,
      773,
      774,
      775
    ],
    "ProjectedImpressions": 78662.75,
    "ProvidedImpressions": 69000.0,
    "SpotCost": 750.0000,
    "StationLegacyCallLetters": "WPIX",
    "Unit": null
  },
  {
    "InventorySource": "Open Market",
    "InventorySourceType": "Open Market",
    "ManifestAudiences": [
      {
        "AudienceId": 31,
        "Impressions": 29600.0,
        "IsReference": true
      },
      {
        "AudienceId": 33,
        "Impressions": 18000.0,
        "IsReference": true
      },
      {
        "AudienceId": 37,
        "Impressions": 18000.0,
        "IsReference": true
      },
      {
        "AudienceId": 39,
        "Impressions": 13700.0,
        "IsReference": true
      },
      {
        "AudienceId": 40,
        "Impressions": 21900.0,
        "IsReference": true
      }
    ],
    "ManifestDayparts": [
      {
        "DaypartId": 32055,
        "Id": 358773,
        "ProgramName": "TWO & A HALF MEN"
      }
    ],
    "ManifestId": 287147,
    "MediaWeekIds": [
      771,
      772,
      773,
      774,
      775
    ],
    "ProjectedImpressions": 67230.55,
    "ProvidedImpressions": 59200.0,
    "SpotCost": 700.0000,
    "StationLegacyCallLetters": "WPIX",
    "Unit": null
  },
  {
    "InventorySource": "Open Market",
    "InventorySourceType": "Open Market",
    "ManifestAudiences": [
      {
        "AudienceId": 31,
        "Impressions": 30200.0,
        "IsReference": true
      },
      {
        "AudienceId": 33,
        "Impressions": 11300.0,
        "IsReference": true
      },
      {
        "AudienceId": 37,
        "Impressions": 11800.0,
        "IsReference": true
      },
      {
        "AudienceId": 39,
        "Impressions": 12100.0,
        "IsReference": true
      },
      {
        "AudienceId": 40,
        "Impressions": 21300.0,
        "IsReference": true
      }
    ],
    "ManifestDayparts": [
      {
        "DaypartId": 32943,
        "Id": 358776,
        "ProgramName": "FRIENDS"
      }
    ],
    "ManifestId": 287150,
    "MediaWeekIds": [
      771,
      772,
      773,
      774,
      775
    ],
    "ProjectedImpressions": 89996.5,
    "ProvidedImpressions": 60400.0,
    "SpotCost": 800.0000,
    "StationLegacyCallLetters": "WPIX",
    "Unit": null
  },
  {
    "InventorySource": "Open Market",
    "InventorySourceType": "Open Market",
    "ManifestAudiences": [
      {
        "AudienceId": 31,
        "Impressions": 21800.0,
        "IsReference": true
      },
      {
        "AudienceId": 33,
        "Impressions": 14100.0,
        "IsReference": true
      },
      {
        "AudienceId": 37,
        "Impressions": 13600.0,
        "IsReference": true
      },
      {
        "AudienceId": 39,
        "Impressions": 12900.0,
        "IsReference": true
      },
      {
        "AudienceId": 40,
        "Impressions": 16900.0,
        "IsReference": true
      }
    ],
    "ManifestDayparts": [
      {
        "DaypartId": 32072,
        "Id": 358777,
        "ProgramName": "FRIENDS"
      }
    ],
    "ManifestId": 287151,
    "MediaWeekIds": [
      771,
      772,
      773,
      774,
      775
    ],
    "ProjectedImpressions": 61366.0,
    "ProvidedImpressions": 43600.0,
    "SpotCost": 750.0000,
    "StationLegacyCallLetters": "WPIX",
    "Unit": null
  },
  {
    "InventorySource": "Open Market",
    "InventorySourceType": "Open Market",
    "ManifestAudiences": [
      {
        "AudienceId": 31,
        "Impressions": 23400.0,
        "IsReference": true
      },
      {
        "AudienceId": 33,
        "Impressions": 11800.0,
        "IsReference": true
      },
      {
        "AudienceId": 37,
        "Impressions": 13300.0,
        "IsReference": true
      },
      {
        "AudienceId": 39,
        "Impressions": 12100.0,
        "IsReference": true
      },
      {
        "AudienceId": 40,
        "Impressions": 16900.0,
        "IsReference": true
      }
    ],
    "ManifestDayparts": [
      {
        "DaypartId": 29548,
        "Id": 358780,
        "ProgramName": "FRIENDS 2X"
      }
    ],
    "ManifestId": 287154,
    "MediaWeekIds": [
      771,
      772,
      773,
      774,
      775
    ],
    "ProjectedImpressions": 80069.375,
    "ProvidedImpressions": 46800.0,
    "SpotCost": 600.0000,
    "StationLegacyCallLetters": "WPIX",
    "Unit": null
  },
  {
    "InventorySource": "Open Market",
    "InventorySourceType": "Open Market",
    "ManifestAudiences": [
      {
        "AudienceId": 31,
        "Impressions": 300.0,
        "IsReference": true
      },
      {
        "AudienceId": 33,
        "Impressions": 100.0,
        "IsReference": true
      },
      {
        "AudienceId": 37,
        "Impressions": 100.0,
        "IsReference": true
      },
      {
        "AudienceId": 39,
        "Impressions": 100.0,
        "IsReference": true
      },
      {
        "AudienceId": 40,
        "Impressions": 300.0,
        "IsReference": true
      }
    ],
    "ManifestDayparts": [
      {
        "DaypartId": 32941,
        "Id": 360318,
        "ProgramName": "TWO & HALF MEN"
      }
    ],
    "ManifestId": 288692,
    "MediaWeekIds": [
      771,
      772,
      773,
      774,
      775
    ],
    "ProjectedImpressions": 175.4,
    "ProvidedImpressions": 600.0,
    "SpotCost": 50.0000,
    "StationLegacyCallLetters": "WPFO",
    "Unit": null
  },
  {
    "InventorySource": "Open Market",
    "InventorySourceType": "Open Market",
    "ManifestAudiences": [
      {
        "AudienceId": 31,
        "Impressions": 300.0,
        "IsReference": true
      },
      {
        "AudienceId": 33,
        "Impressions": 100.0,
        "IsReference": true
      },
      {
        "AudienceId": 37,
        "Impressions": 300.0,
        "IsReference": true
      },
      {
        "AudienceId": 39,
        "Impressions": 300.0,
        "IsReference": true
      },
      {
        "AudienceId": 40,
        "Impressions": 400.0,
        "IsReference": true
      }
    ],
    "ManifestDayparts": [
      {
        "DaypartId": 29555,
        "Id": 360319,
        "ProgramName": "2.5 MEN X2"
      }
    ],
    "ManifestId": 288693,
    "MediaWeekIds": [
      771,
      772,
      773,
      774,
      775
    ],
    "ProjectedImpressions": 434.625,
    "ProvidedImpressions": 600.0,
    "SpotCost": 12.0000,
    "StationLegacyCallLetters": "WPFO",
    "Unit": null
  },
  {
    "InventorySource": "Open Market",
    "InventorySourceType": "Open Market",
    "ManifestAudiences": [
      {
        "AudienceId": 31,
        "Impressions": 200.0,
        "IsReference": true
      },
      {
        "AudienceId": 33,
        "Impressions": 100.0,
        "IsReference": true
      },
      {
        "AudienceId": 37,
        "Impressions": 100.0,
        "IsReference": true
      },
      {
        "AudienceId": 39,
        "Impressions": 100.0,
        "IsReference": true
      },
      {
        "AudienceId": 40,
        "Impressions": 200.0,
        "IsReference": true
      }
    ],
    "ManifestDayparts": [
      {
        "DaypartId": 29548,
        "Id": 360320,
        "ProgramName": "SEINFELD X2"
      }
    ],
    "ManifestId": 288694,
    "MediaWeekIds": [
      771,
      772,
      773,
      774,
      775
    ],
    "ProjectedImpressions": 580.0,
    "ProvidedImpressions": 400.0,
    "SpotCost": 12.0000,
    "StationLegacyCallLetters": "WPFO",
    "Unit": null
  },
  {
    "InventorySource": "Open Market",
    "InventorySourceType": "Open Market",
    "ManifestAudiences": [
      {
        "AudienceId": 31,
        "Impressions": 200.0,
        "IsReference": true
      },
      {
        "AudienceId": 33,
        "Impressions": 100.0,
        "IsReference": true
      },
      {
        "AudienceId": 37,
        "Impressions": 100.0,
        "IsReference": true
      },
      {
        "AudienceId": 39,
        "Impressions": 100.0,
        "IsReference": true
      },
      {
        "AudienceId": 40,
        "Impressions": 200.0,
        "IsReference": true
      }
    ],
    "ManifestDayparts": [
      {
        "DaypartId": 32055,
        "Id": 360321,
        "ProgramName": "2 BROKE GIRLS"
      }
    ],
    "ManifestId": 288695,
    "MediaWeekIds": [
      771,
      772,
      773,
      774,
      775
    ],
    "ProjectedImpressions": 117.4,
    "ProvidedImpressions": 400.0,
    "SpotCost": 12.0000,
    "StationLegacyCallLetters": "WPFO",
    "Unit": null
  }
]